composer.json

```diff
@@ -14,8 +14,11 @@
   "require-dev": {
     "dealerdirect/phpcodesniffer-composer-installer": "^0.7",
     "phpcompatibility/php-compatibility": "^9.3",
+    "phpstan/extension-installer": "^1.3.0",
+    "phpstan/phpstan": "^1.10",
     "phpunit/phpunit": "^4|^5|^6|^7|^8|^9",
     "squizlabs/php_codesniffer": "^3.5",
+    "szepeviktor/phpstan-wordpress": "^1.3.0",
```

Comment:
I can also highly recommend https://packagist.org/packages/swissspidy/phpstan-no-private to catch calls of seemingly private functions.

Plus https://packagist.org/packages/phpstan/phpstan-phpunit and https://packagist.org/packages/php-stubs/wordpress-tests-stubs to run PHPStan against tests as well. Believe me, it really improves the quality of the tests!

https://packagist.org/packages/phpstan/phpstan-deprecation-rules is also nice to prevent usage of deprecated functions.

Reply:
> I can also highly recommend https://packagist.org/packages/swissspidy/phpstan-no-private to catch calls of seemingly private functions.

I love this package, but I wonder if this is the best for for Performance Lab considering that the code in question will end up in core and thus it is expected that many calls to internal/private functions will be made? As it stands right now...

<details><summary>There are 73 instances of this</summary>
<p>

```
 ------ ------------------------------------------------------- 
  Line   modules/database/sqlite/activate.php                   
 ------ ------------------------------------------------------- 
  71     Call to private/internal function wp_set_wpdb_vars().  
  111    Call to private/internal function wp_set_wpdb_vars().  
 ------ ------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------ 
  Line   modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-alter-query.php                 
 ------ ------------------------------------------------------------------------------------------------ 
  53     Call to private/internal method command_tokenizer() of class Perflab_SQLite_Alter_Query.        
  65     Call to private/internal method handle_single_command() of class Perflab_SQLite_Alter_Query.    
  69     Call to private/internal method handle_add_primary_key() of class Perflab_SQLite_Alter_Query.   
  73     Call to private/internal method handle_drop_primary_key() of class Perflab_SQLite_Alter_Query.  
  77     Call to private/internal method handle_modify_command() of class Perflab_SQLite_Alter_Query.    
  81     Call to private/internal method handle_change_command() of class Perflab_SQLite_Alter_Query.    
  85     Call to private/internal method handle_alter_command() of class Perflab_SQLite_Alter_Query.     
  261    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
  367    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
  421    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
  483    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
  500    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
  513    Call to private/internal method convert_field_types() of class Perflab_SQLite_Alter_Query.      
 ------ ------------------------------------------------------------------------------------------------ 

 ------ ----------------------------------------------------------------------------------------------- 
  Line   modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-create-query.php               
 ------ ----------------------------------------------------------------------------------------------- 
  76     Call to private/internal method strip_backticks() of class Perflab_SQLite_Create_Query.        
  77     Call to private/internal method quote_illegal_field() of class Perflab_SQLite_Create_Query.    
  78     Call to private/internal method get_table_name() of class Perflab_SQLite_Create_Query.         
  79     Call to private/internal method rewrite_comments() of class Perflab_SQLite_Create_Query.       
  80     Call to private/internal method rewrite_field_types() of class Perflab_SQLite_Create_Query.    
  81     Call to private/internal method rewrite_character_set() of class Perflab_SQLite_Create_Query.  
  82     Call to private/internal method rewrite_engine_info() of class Perflab_SQLite_Create_Query.    
  83     Call to private/internal method rewrite_unsigned() of class Perflab_SQLite_Create_Query.       
  84     Call to private/internal method rewrite_autoincrement() of class Perflab_SQLite_Create_Query.  
  85     Call to private/internal method rewrite_primary_key() of class Perflab_SQLite_Create_Query.    
  86     Call to private/internal method rewrite_foreign_key() of class Perflab_SQLite_Create_Query.    
  87     Call to private/internal method rewrite_unique_key() of class Perflab_SQLite_Create_Query.     
  88     Call to private/internal method rewrite_enum() of class Perflab_SQLite_Create_Query.           
  89     Call to private/internal method rewrite_set() of class Perflab_SQLite_Create_Query.            
  90     Call to private/internal method rewrite_key() of class Perflab_SQLite_Create_Query.            
  91     Call to private/internal method add_if_not_exists() of class Perflab_SQLite_Create_Query.      
  93     Call to private/internal method post_process() of class Perflab_SQLite_Create_Query.           
 ------ ----------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------- 
  Line   modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-pdo-driver.php                      
 ------ ---------------------------------------------------------------------------------------------------- 
  81     Call to private/internal method parse_query() of class Perflab_SQLite_PDO_Driver.                   
  84     Call to private/internal method handle_truncate_query() of class Perflab_SQLite_PDO_Driver.         
  88     Call to private/internal method handle_alter_query() of class Perflab_SQLite_PDO_Driver.            
  92     Call to private/internal method handle_create_query() of class Perflab_SQLite_PDO_Driver.           
  97     Call to private/internal method handle_describe_query() of class Perflab_SQLite_PDO_Driver.         
  101    Call to private/internal method handle_show_query() of class Perflab_SQLite_PDO_Driver.             
  105    Call to private/internal method handle_show_columns_query() of class Perflab_SQLite_PDO_Driver.     
  109    Call to private/internal method handle_show_index() of class Perflab_SQLite_PDO_Driver.             
  115    Call to private/internal method rewrite_date_sub() of class Perflab_SQLite_PDO_Driver.              
  116    Call to private/internal method delete_index_hints() of class Perflab_SQLite_PDO_Driver.            
  117    Call to private/internal method rewrite_regexp() of class Perflab_SQLite_PDO_Driver.                
  119    Call to private/internal method fix_date_quoting() of class Perflab_SQLite_PDO_Driver.              
  120    Call to private/internal method rewrite_between() of class Perflab_SQLite_PDO_Driver.               
  121    Call to private/internal method handle_orderby_field() of class Perflab_SQLite_PDO_Driver.          
  126    Call to private/internal method execute_duplicate_key_update() of class Perflab_SQLite_PDO_Driver.  
  127    Call to private/internal method rewrite_insert_ignore() of class Perflab_SQLite_PDO_Driver.         
  128    Call to private/internal method rewrite_regexp() of class Perflab_SQLite_PDO_Driver.                
  129    Call to private/internal method fix_date_quoting() of class Perflab_SQLite_PDO_Driver.              
  134    Call to private/internal method rewrite_update_ignore() of class Perflab_SQLite_PDO_Driver.         
  136    Call to private/internal method rewrite_limit_usage() of class Perflab_SQLite_PDO_Driver.           
  137    Call to private/internal method rewrite_order_by_usage() of class Perflab_SQLite_PDO_Driver.        
  138    Call to private/internal method rewrite_regexp() of class Perflab_SQLite_PDO_Driver.                
  139    Call to private/internal method rewrite_between() of class Perflab_SQLite_PDO_Driver.               
  144    Call to private/internal method rewrite_limit_usage() of class Perflab_SQLite_PDO_Driver.           
  145    Call to private/internal method rewrite_order_by_usage() of class Perflab_SQLite_PDO_Driver.        
  146    Call to private/internal method rewrite_date_sub() of class Perflab_SQLite_PDO_Driver.              
  147    Call to private/internal method rewrite_regexp() of class Perflab_SQLite_PDO_Driver.                
  148    Call to private/internal method delete_workaround() of class Perflab_SQLite_PDO_Driver.             
  153    Call to private/internal method rewrite_date_sub() of class Perflab_SQLite_PDO_Driver.              
  154    Call to private/internal method rewrite_regexp() of class Perflab_SQLite_PDO_Driver.                
  158    Call to private/internal method rewrite_optimize() of class Perflab_SQLite_PDO_Driver.              
  168    Call to private/internal method return_true() of class Perflab_SQLite_PDO_Driver.                   
 ------ ---------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------- 
  Line   modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-pdo-engine.php                         
 ------ ------------------------------------------------------------------------------------------------------- 
  1260   Call to private/internal method convert_to_columns_object() of class Perflab_SQLite_PDO_Engine.        
  1262   Call to private/internal method convert_to_index_object() of class Perflab_SQLite_PDO_Engine.          
  1264   Call to private/internal method convert_result_check_or_analyze() of class Perflab_SQLite_PDO_Engine.  
 ------ ------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------- 
  Line   modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-pdo-user-defined-functions.php         
 ------ ------------------------------------------------------------------------------------------------------- 
  324    Call to private/internal method derive_interval() of class Perflab_SQLite_PDO_User_Defined_Functions.  
  356    Call to private/internal method derive_interval() of class Perflab_SQLite_PDO_User_Defined_Functions.  
 ------ ------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------ 
  Line   modules/images/dominant-color-images/hooks.php                                
 ------ ------------------------------------------------------------------------------ 
  19     Call to private/internal function _dominant_color_get_dominant_color_data().  
 ------ ------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------ 
  Line   modules/images/webp-uploads/helper.php                                              
 ------ ------------------------------------------------------------------------------------ 
  222    Call to private/internal function webp_uploads_generate_additional_image_source().  
 ------ ------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------ 
  Line   modules/images/webp-uploads/hooks.php                                               
 ------ ------------------------------------------------------------------------------------ 
  82     Call to private/internal function webp_uploads_generate_additional_image_source().  
  126    Call to private/internal function _wp_image_meta_replace_original().                
 ------ ------------------------------------------------------------------------------------ 
```

</p>
</details>

Reply:
> Plus https://packagist.org/packages/phpstan/phpstan-phpunit

Thanks for that. Added in 916dd25484ea42fe16356e80816e018fa1fad8ac, fixing 2 found issues in 5e498d1b28b9f74ab25e9b8037032c67b85f64c5.

Reply:
> and https://packagist.org/packages/php-stubs/wordpress-tests-stubs

Do we need the test subs since we're already installing `wp-phpunit/wp-phpunit`? Would this essentially be an alternative to:

https://github.com/WordPress/performance/blob/5e498d1b28b9f74ab25e9b8037032c67b85f64c5/phpstan.neon.dist#L14-L15

Reply:
> https://packagist.org/packages/phpstan/phpstan-deprecation-rules is also nice to prevent usage of deprecated functions.

Added in 4554f6af2b7aa9343d57eb16b8e33fc214833a14, issues fixed in bd1c3783570b23fd3213d25b189415bd7ac3936c.

Reply:
> I love this package, but I wonder if this is the best for for Performance Lab considering that the code in question will end up in core and thus it is expected that many calls to internal/private functions will be made? As it stands right now...

Good point! 😃 

I think it can't hurt to use it anyway, since not everything will land in core. Places where usage is OK can be ignored in the PHPStan config accordingly. But maybe in another PR :-)

All the `Perflab_SQLite_` reports are actually false positives! I'll get this fixed in a new release.
They are not properly documented though. A `private function ...()` method does not need `@access private` annotations. That's against the coding standards.

> Do we need the test subs since we're already installing `wp-phpunit/wp-phpunit`? Would this essentially be an alternative to:

Probably not in that case. Counter question: why is `wp-phpunit` even used here when all tests are running in wp-env anyway?

Reply:
Agreed, also if our code here follows PHPStan, I don't think it would violate anything that core requires - we'd effectively have stricter requirements than core, which I think should work fine as long as it's in that direction and not the opposite :)

> why is `wp-phpunit` even used here when all tests are running in wp-env anyway?

I believe that is for additional flexibility. We don't want to _dictate_ anyone to use `wp-env`, it's just the easy way to go. If folks want to e.g. run a DB on their own machine, the use of `wp-phpunit` makes that cleaner and decouples from having to have the entire WP core dev environment.